cpp/src/arrow/compute/expression.cc

```diff
@@ -874,7 +874,7 @@ Result<Expression> FoldConstants(Expression expr) {
           }
         }
 
-        if (call->function_name == "and_kleene") {
+        if (call->function_name == "and_kleene" || call->function_name == "and") {
```

Comment:
These are not correct. The SQL logical operators `AND` and `OR` correspond to `and_kleene` and `or_kleene`, whereas `and(false, null) -> null`.

I'm not sure the non-Kleene logical functions are used enough to warrant special case optimizations

Reply:
> These are not correct. The SQL logical operators `AND` and `OR` correspond to `and_kleene` and `or_kleene`, whereas `and(false, null) -> null`.

Yes, you're right. `OR` is same  with `or_kleene`.

> I'm not sure the non-Kleene logical functions are used enough to warrant special case optimizations

Yes, SQL logical operators should do these things by their own sql-optimizer with sql-standard grammar. And the kleene logical functions seems enough for other compute operators?

Reply:
> Yes, SQL logical operators should do these things by their own sql-optimizer with sql-standard grammar. And the kleene logical functions seems enough for other compute operators?

I'm confused; SQL's AND == `and_kleene`. `and` does not have a SQL equivalent to my knowledge.

Reply:
Oh, my fault. In SQL scenarios, Kleene-related logical operators are commonly used. Optimizing for Kleene operators is sufficient in most cases, and it may not be necessary to optimize for non-Kleene operators. This is not a universally applicable scenario within the expression system.

Reply:
Okay, then maybe we should close this PR if we don't need to optimize for non-Kleene operators.

Reply:
Okay!